### PR TITLE
Update post author block functionality and visual parity

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
## Description

Update the post author block functionality as issued in #21087.

This introduces:
* <s>Author font size</s>
* <s>Byline font size</s>
* <s>Bio font size</s>
* __experimentalFontSize support
* Relative sizing for byline and bio fields.
* Remaining Rich Text capabilities for the byline.

This also fixes issues with:
* Text alignment
* Font and Background Colors
* Default attributes
* Overall parity between editors and front-end

editing-post-author

### A note on front-end parity:
Preset font sizes ( 'Normal', 'Large', etc. ) run on class names that may have styles added by themes.  For example, TwentyTwenty adds styles such as:
```
.entry-content .has-large-font-size {
    font-size: 1.25em;
    line-height: 1.4;
}
```
This overwrites the standard definition for large text on the front end and causes it to be a different size than the expected 36px as is shown in the editor.  Im not sure what we can/should do about this, as if we force parity on this we disable the themes ability to have control over preset sizing.

## How has this been tested?

Tested on local docker environment.
Test various settings, save, and verify visual parity between site editor, post editor, and front end.

## Screenshots 
Screen Shot 2020-06-04 at 7 05 32 PM

Screen Shot 2020-06-04 at 8 16 14 PM

## Types of changes

Bug fix (non-breaking change which fixes an issue)
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by epiqueras — COMMENTED** [review]
Can we use `FONT_SIZE_SUPPORT_KEY = '__experimentalFontSize'` here?

**Comment by Addison-Stavlo** [comment]
> Can we use FONT_SIZE_SUPPORT_KEY = '__experimentalFontSize' here?

Im not sure I follow.  Use this where / for what?  Is this some alternative to `withFontSizes` ?

**Comment by Addison-Stavlo** [comment]
Oh would that be to support the global styles set fontSize?

This block is odd that with name, byline, and bio, there are 3 distinct sections that should support being different sizes.  Overwriting them all with the same global styles size would severely limit the styling capabilities of the block.

**Comment by epiqueras** [comment]
It's a new way of doing font sizes.

Perhaps the secondary text should be sized relative to the "master" font size? Otherwise, it sounds like they should be their own child blocks.

**Comment by Addison-Stavlo** [comment]
> It's a new way of doing font sizes.

Ah gotcha.  Im not familiar with it atm, but can take a look tomorrow.  This would require breaking the 3 sections up into child blocks and using this file?
https://github.com/WordPress/gutenberg/blob/32af45067b046e9449789358ad21482d1d3393b5/packages/block-editor/src/hooks/font-size.js#L99

The `withFontSizes` lets us set up all 3 without the need to create child blocks, would this not make more sense currently?  I just kind of took off with the font size tool this had been initialized with.

**Comment by epiqueras** [comment]
The Heading block is an example of how to use it.

We need `__experimentalFontSize` to hook into global styles. For now, I recommend the various font sizes are set relative to the font size set for the entire block. Later, if people require it, we can look at creating child blocks or using `__experimentalSelector` to overwrite specific font sizes.

**Comment by Addison-Stavlo** [comment]
Interesting!  Thanks for pointing me in the right direction.

I have updated this to use the `__experimentalFontSize` instead, and converted the byline and bio texts to be sized relative to that standard fontSize (0.5em and 0.7em respectively).  There may be more appropriate values to set for these. 🤔 

Im still a bit uncertain regarding global styles:  would this then inherit what is set for global styles by default, and once we specifically save a font size on the block it will use that instead?

**Comment by epiqueras** [comment]
>Im still a bit uncertain regarding global styles: would this then inherit what is set for global styles by default, and once we specifically save a font size on the block it will use that instead?

Yes

File: packages/block-library/src/post-author/edit.js [hunk 1]
```diff
@@ -72,15 +83,15 @@ function PostAuthorEdit( {
...
+					fontSize: contrastCheckFontSize,
 				},
 			],
 			colorDetector: { targetRef: ref },
 			colorPanelProps: {
 				initialOpen: true,
 			},
 		},
-		[ fontSize.size ]
+		[ contrastCheckFontSize ]
```
Comment (epiqueras) [inline packages/block-library/src/post-author/edit.js:94]: We need to change this to `__experimentalUseColors`.

File: packages/block-library/src/post-author/edit.js [hunk 2]
```diff
@@ -205,12 +202,6 @@ function PostAuthorEdit( {
 									className="wp-block-post-author__byline"
 									multiline={ false }
 									placeholder={ __( 'Write byline …' ) }
-									withoutInteractiveFormatting
-									allowedFormats={ [
-										'core/bold',
-										'core/italic',
-										'core/strikethrough',
-									] }
```
Comment (epiqueras) [inline packages/block-library/src/post-author/edit.js:213]: What's the reasoning behind this change?

File: packages/block-library/src/post-author/editor.scss [hunk 3]
```diff
@@ -1,12 +1,12 @@
-.block-editor__container .wp-block-post-author {
+.block-editor__container .wp-block-post-author,
+.edit-site-visual-editor .wp-block-post-author {
```
Comment (epiqueras) [inline packages/block-library/src/post-author/editor.scss:2]: Why is this using `block-editor__container`. I don't think that selector is used anywhere.

File: packages/block-library/src/post-author/index.php [hunk 4]
```diff
@@ -40,12 +40,6 @@ function post_author_build_css_colors( $attributes ) {
 	$has_named_background_color  = array_key_exists( 'backgroundColor', $attributes );
 	$has_custom_background_color = array_key_exists( 'customBackgroundColor', $attributes );
 
-	// If has background color.
-	if ( $has_custom_background_color || $has_named_background_color ) {
-		// Add has-background-color class.
-		$colors['css_classes'][] = 'has-background-color';
-	}
```
Comment (epiqueras) [inline packages/block-library/src/post-author/index.php:47]: Why is this removed?

File: packages/block-library/src/post-author/index.php [hunk 5]
```diff
@@ -72,16 +66,21 @@ function post_author_build_css_font_sizes( $attributes ) {
...
 		$font_sizes['css_classes'][] = sprintf( 'has-%s-font-size', $attributes['fontSize'] );
 	} elseif ( $has_custom_font_size ) {
 		// Add the custom font size inline style.
-		$font_sizes['inline_styles'] = sprintf( 'font-size: %spx;', $attributes['customFontSize'] );
+		$font_sizes['inline_styles'] = sprintf( 'font-size: %spx;', $attributes['style']['typography']['fontSize'] );
 	}
 
+	$has_named_bio_font_size  = array_key_exists( 'bioFontSize', $attributes );
+	$has_custom_bio_font_size = array_key_exists( 'customBioFontSize', $attributes );
+
```
Comment (epiqueras) [inline packages/block-library/src/post-author/index.php:83]: I don't think this is used.

File: packages/block-library/src/post-author/index.php [hunk 6]
```diff
@@ -117,7 +116,7 @@ function render_block_core_post_author( $attributes, $content, $block ) {
 		array( 'wp-block-post-author' ),
 		isset( $attributes['className'] ) ? array( $attributes['className'] ) : array(),
 		isset( $attributes['itemsJustification'] ) ? array( 'items-justified-' . $attributes['itemsJustification'] ) : array(),
-		isset( $attributes['align'] ) ? array( 'align' . $attributes['align'] ) : array()
+		isset( $attributes['align'] ) ? array( 'has-text-align-' . $attributes['align'] ) : array()
```
Comment (epiqueras) [inline packages/block-library/src/post-author/index.php:119]: Is this change related to this PR?

File: packages/block-library/src/post-author/edit.js (hunk 1 above)
Comment (epiqueras) [reply packages/block-library/src/post-author/edit.js:94]: (In a different PR.)

**Review by epiqueras — CHANGES_REQUESTED** [review]

File: packages/block-library/src/post-author/edit.js (hunk 2 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/post-author/edit.js:213]: I figured that if the goal was to have parity with paragraph block's functionalities, that the editable text should be able to have links and the remaining text formatting options.  Removing these removes the restrictions and enables them.

File: packages/block-library/src/post-author/index.php (hunk 4 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/post-author/index.php:47]: It causes themes to overwrite the font color that gets set and saved.  Themes apply styles to this to selector to set color on the font to ensure it is still readable.  If we are also controlling the font color, then it is unnecessary as well.  Considering it more, we should re-apply this class name, but only when a font color is not set.

File: packages/block-library/src/post-author/index.php (hunk 6 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/post-author/index.php:119]: That 'align' attribute adjusts the blocks margin on the page.  This attribute corresponds with the text alignment setting so I changed the class name to appropriately reflect that.  

Previously selecting 'align-right' for text alignment would then the block would move to the right of the page, with its text alignment unchanged.

File: packages/block-library/src/post-author/editor.scss (hunk 3 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/post-author/editor.scss:2]: Yeah, `.block-editor__container .wp-block-post-author` was the only selector that was there previously.  Its part of the post editor, but I don't know what it was the class originally put here.  To be safe I just added the `.edit-site-visual-editor .wp-block-post-author` selector as well so the styles would be in the site editor as well.

File: packages/block-library/src/post-author/index.php (hunk 4 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/post-author/index.php:47]: > Considering it more, we should re-apply this class name, but only when a font color is not set.

Or should we?  In that case if someone adds this block and changes the background to a very light color (without changing any font color), a theme would still then force their font color to be really light as well.

Comment (Addison-Stavlo) [reply packages/block-library/src/post-author/index.php:47]: Maybe the best solution is to add this class back in and move the font color styles that get set to inline styles.

File: packages/block-library/src/post-author/edit.js (hunk 2 above)
Comment (epiqueras) [reply packages/block-library/src/post-author/edit.js:213]: I think these specific restrictions were put in place because other formatting options don't make sense in the context of a byline.

File: packages/block-library/src/post-author/editor.scss (hunk 3 above)
Comment (epiqueras) [reply packages/block-library/src/post-author/editor.scss:2]: I don't even see it anywhere in the DOM. We should just use `.wp-block-post-author`.

File: packages/block-library/src/post-author/index.php (hunk 4 above)
Comment (epiqueras) [reply packages/block-library/src/post-author/index.php:47]: We should mirror what the other core blocks are doing.

File: packages/block-library/src/post-author/index.php (hunk 6 above)
Comment (epiqueras) [reply packages/block-library/src/post-author/index.php:116]: So this was a bug.

File: packages/block-library/src/post-author/index.php (hunk 4 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/post-author/index.php]: Ok, I re-added that class and moved the font color classes/styles into the content div instead.

File: packages/block-library/src/post-author/edit.js (hunk 2 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/post-author/edit.js:213]: Why restrict it thought?  If a user wants a word in their byline to be a link, then all of a sudden a link makes sense in the context of a byline for them.  A similar argument for subscript/superscript, etc.

File: packages/block-library/src/post-author/editor.scss (hunk 3 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/post-author/editor.scss]: > I don't even see it anywhere in the DOM. 

Odd, for me its a child of the .block-editor:
Screen Shot 2020-06-05 at 6 47 22 PM

Either way, I think just using `.wp-block-post-author` makes sense.  Updated.

**Review by epiqueras — APPROVED** [review]
